packages/core/saved-objects/core-saved-objects-migration-server-internal/src/zdt/migration_state_action_machine.ts

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { errors as EsErrors } from '@elastic/elasticsearch';
+import type { Logger } from '@kbn/logging';
+import {
+  getErrorMessage,
+  getRequestDebugMeta,
+} from '@kbn/core-elasticsearch-client-server-internal';
+import { logStateTransition, logActionResponse } from '../common/utils';
+import { type Next, stateActionMachine } from '../state_action_machine';
+import { cleanup } from '../migrations_state_machine_cleanup';
+import type { State } from './state';
+import type { MigratorContext } from './context';
+
+/**
+ * A specialized migrations-specific state-action machine that:
+ *  - logs messages in state.logs
+ *  - logs state transitions
+ *  - logs action responses
+ *  - resolves if the final state is DONE
+ *  - rejects if the final state is FATAL
+ *  - catches and logs exceptions and then rejects with a migrations specific error
+ */
+export async function migrationStateActionMachine({
+  initialState,
+  context,
+  next,
+  model,
+  logger,
+}: {
+  initialState: State;
+  context: MigratorContext;
+  next: Next<State>;
+  model: (state: State, res: any, context: MigratorContext) => State;
+  logger: Logger;
+}) {
```

Comment:
So, as much as I tried to reuse every possible components between the 2 algos, I was forced to duplicate the `migrationStateActionMachine`, for the simple reason that the model signature isn't compatible because of the introduction of the `context` as a third parameter (and also as an additional input to the whole function).

I think it's fine, it's ~= 150 lines and we could even factorize the whole error handling logic by extracting it, which would remove most of the duplication.